source/tests/system/visa_session_tests.cpp

```diff
@@ -0,0 +1,113 @@
+#include "device_server.h"
+#include "tests/utilities/test_helpers.h"
+#include "visa/visa_client.h"
+
+namespace ni {
+namespace tests {
+namespace system {
+
+namespace visa = visa_grpc;
+
+const int kInvalidRsrc = -1073807343;
+const int kInvalidVisaSessionWarning = 1073676418;
+const char* kVisaErrorInstrumentDescriptorNotFoundMessage = "Insufficient location information or the device or resource is not present in the system.";
+const char* kInstrumentDescriptor = "TCPIP::www.ni.com::80::SOCKET";
+const char* kVisaTestSession = "SessionName";
+const char* kVisaTestInvalidInstrumentDescriptor = "";
+
+class VisaSessionTest : public ::testing::Test {
+ protected:
+  VisaSessionTest()
+      : device_server_(DeviceServerInterface::Singleton()),
+        visa_stub_(visa::Visa::NewStub(device_server_->InProcessChannel()))
+  {
+    device_server_->ResetServer();
+  }
+
+  virtual ~VisaSessionTest() {}
+
+  std::unique_ptr<visa::Visa::Stub>& GetStub()
+  {
+    return visa_stub_;
+  }
+
+  ::grpc::Status call_open(const char* instrument_descriptor, visa::LockState access_mode, const char* session_name, uint32_t open_timeout, visa::OpenResponse* response)
+  {
+    ::grpc::ClientContext context;
+    visa::OpenRequest request;
+    request.set_instrument_descriptor(instrument_descriptor);
+    request.set_access_mode(access_mode);
+    request.set_session_name(session_name);
+    request.set_open_timeout(open_timeout);
+
+    ::grpc::Status status = GetStub()->Open(&context, request, response);
+    nidevice_grpc::experimental::client::raise_if_error(status, context);
+    return status;
+  }
+
+ private:
+  DeviceServerInterface* device_server_;
+  std::unique_ptr<visa::Visa::Stub> visa_stub_;
+};
+
+TEST_F(VisaSessionTest, OpenSessionWithInstrumentDescriptor_CreatesDriverSession)
+{
+  visa::OpenResponse response;
+  ::grpc::Status status = call_open(kInstrumentDescriptor, visa::LOCK_STATE_NO_LOCK, kVisaTestSession, 0, &response);
+
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(0, response.status());
+  EXPECT_NE("", response.vi().name());
```

Comment:
If we request a specific session name, this should test equality against that. (In the next test where you specify an empty string, this check makes sense there.)

Reply:
Commit 5